### PR TITLE
Entropy method for calibration-based quantization

### PR DESCRIPTION
**Pull request by chilo-ms**
**Description**: Add entropy method for calibration-based quantization.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/python/tools/quantization/calibrate.py [hunk 1]
```diff
@@ -226,12 +227,258 @@ def compute_range(self):
...
+        self.intermediate_outputs = []
+        self.calibrate_tensors_range = None
+        self.num_model_outputs = len(self.model.graph.output)
+        self.model_original_outputs = set(output.name for output in self.model.graph.output)
+        self.collector = None
+
+    def augment_graph(self):
+        '''
+        Adds ReduceMin and ReduceMax nodes to all quantization_candidates op type nodes in
+        model and ensures their outputs are stored as part of the graph output
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/calibrate.py:247]: this description is not accurate.

File: onnxruntime/python/tools/quantization/calibrate.py [hunk 2]
```diff
@@ -226,12 +227,258 @@ def compute_range(self):
...
+
+        for node in model.graph.node:
+            if len(self.op_types_to_calibrate) == 0 or node.op_type in self.op_types_to_calibrate:
+                for tensor_name in itertools.chain(node.input, node.output):
+                    if tensor_name in value_infos.keys():
+                        vi = value_infos[tensor_name]
+                        if vi.type.HasField('tensor_type') and (
+                                vi.type.tensor_type.elem_type in tensor_type_to_calibrate) and (
+                                    tensor_name not in initializer):
+                            tensors_to_calibrate.add(tensor_name)
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/calibrate.py:271]: this is same with the MaxMin one. Could you extract the code and make it reusable

Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py]: Made it reusable.

File: onnxruntime/python/tools/quantization/calibrate.py [hunk 3]
```diff
@@ -226,12 +239,242 @@ def compute_range(self):
...
+            self.collector = HistogramCollector()
+        self.collector.collect(clean_merged_dict)
+
+    def compute_range(self):
+        ''' 
+        Compute the min-max range of tensor
+        :return: dictionary mapping: {added node names: (ReduceMin, ReduceMax) pairs }
+        '''
+        if not self.collector:
+            print("No collector created and can't generate calibration data.")
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/calibrate.py:316]: >print [](start = 12, length = 5)

could you please use logging to keep consistent

File: onnxruntime/python/tools/quantization/calibrate.py [hunk 4]
```diff
@@ -226,12 +239,242 @@ def compute_range(self):
...
+    """
+
+    @abc.abstractmethod
+    def collect(self, name_to_arr):
+        """
+        Generate informative data based on given data.
+            name_to_arr : dict 
+                tensor name to NDArray data 
+        """
+        raise NotImplementedError
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/calibrate.py:334]: Does this abstract class need a get_optimal_collection_result method? Is this abstract class nessarary?

File: onnxruntime/python/tools/quantization/calibrate.py [hunk 5]
```diff
@@ -226,12 +239,242 @@ def compute_range(self):
...
+                new_threshold = half_increased_bins * old_stride + old_threshold
+                hist, hist_edges = np.histogram(data_arr, new_num_bins, range=(-new_threshold, new_threshold))
+                hist[half_increased_bins:new_num_bins-half_increased_bins] += old_hist
+            return (hist, hist_edges, min(old_min, new_min), max(old_max, new_max), new_threshold)
+
+    def get_optimal_collection_result(self):
+        return self.get_optimal_thresholds(self.histogram_dict, self.num_quantized_bins)
+
+
+    def get_optimal_thresholds(self, histogram_dict, num_quantized_bins=128):
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/calibrate.py:396]: Is it more simple to merge them together?

**Comment by yufenglee** [comment]
Could you add the entropy calibration as one option in the quantize_static?

File: onnxruntime/python/tools/quantization/calibrate.py (hunk 3 above)
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:316]: Done

File: onnxruntime/python/tools/quantization/calibrate.py (hunk 4 above)
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:334]: Besides histogram collection, it's possible that later other collection approaches will be introduced.
It's better off to have this abstract class to make sure all the collection approaches are consistent.

get_optimal_collection_result method added.

File: onnxruntime/python/tools/quantization/calibrate.py (hunk 5 above)
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:396]: I've merged the them.

File: onnxruntime/python/tools/quantization/calibrate.py (hunk 1 above)
Comment (jywu-msft) [reply onnxruntime/python/tools/quantization/calibrate.py]: update this description?

File: onnxruntime/python/tools/quantization/E2E_example_model/object_detection/trt/yolov3/evaluate.py [hunk 6]
```diff
@@ -10,6 +10,8 @@
 from onnxruntime.quantization.calibrate import CalibrationDataReader
 import numpy as np
 
+import torch
```
Comment (jywu-msft) [inline onnxruntime/python/tools/quantization/E2E_example_model/object_detection/trt/yolov3/evaluate.py:13]: are we able to implement this without torch dependency? 
looks like it's mainly for NMS and a couple other operations? 
add it as a TODO?

**Review by jywu-msft — DISMISSED** [review]

**Comment by chilo-ms** [comment]
> Could you add the entropy calibration as one option in the quantize_static?

entropy/minmax calibration options added.

File: onnxruntime/python/tools/quantization/E2E_example_model/object_detection/trt/yolov3/evaluate.py (hunk 6 above)
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/E2E_example_model/object_detection/trt/yolov3/evaluate.py:13]: I think we can implement operations without torch, but it takes time to investigate and implement.
Will do it later.

**Review by jywu-msft — APPROVED** [review]

**Pull request merged**